src/libcrun/net_device.c

```diff
@@ -0,0 +1,479 @@
+
+  ret = send_request (sock, req, err);
+  if (UNLIKELY (ret < 0))
+    return ret;
+
+  return wait_for_ack (sock, seq, buffer, buffer_size, err);
+}
+
+static int
+setup_network_device_in_ns_helper (char *buffer, size_t buffer_size, int netns_fd,
+                                   const char *newifname, const char *ips, libcrun_error_t *err)
+{
+  cleanup_close int sock_in_ns = -1;
+  int new_ifindex;
+  int ret;
+
+  ret = setns (netns_fd, CLONE_NEWNET);
+  if (UNLIKELY (ret < 0))
+    return crun_make_error (err, errno, "change network namespace");
+
+  sock_in_ns = open_netlink_fd (err);
+  if (sock_in_ns < 0)
+    return sock_in_ns;
+
+  /* we could ask for a specific index with IFLA_NEW_IFINDEX, and apparently the kernel tries anyway to
+     reuse the existing one, but asking for a specific index could cause conflicts if the
+     target network namespace already exists, so avoid doing it and lookup the device again.  */
+  new_ifindex = name_to_index (sock_in_ns, newifname, buffer, buffer_size, err);
+  if (UNLIKELY (new_ifindex < 0))
+    return new_ifindex;
+
+  ret = configure_ips (sock_in_ns, new_ifindex, buffer, buffer_size, ips, err);
+  if (ret < 0)
+    return ret;
+
+  return request_set_link_state_and_wait (sock_in_ns, buffer, buffer_size, new_ifindex, true, err);
+}
+
+static int
+do_move_link_to_ns_and_wait (int sock, char *buffer, size_t buffer_size, int ifindex, int netns_fd, const char *newifname, libcrun_error_t *err)
+{
+  struct nl_req *req = (struct nl_req *) buffer;
+  uint32_t seq;
+  int ret;
+
+  seq = reset_request (req, RTM_NEWLINK, NLM_F_REQUEST | NLM_F_ACK, sizeof (struct ifinfomsg));
+  req->ifi.ifi_family = AF_UNSPEC;
+  req->ifi.ifi_index = ifindex;
+
+  ret = append_rtattr (&req->nlh, buffer_size, IFLA_NET_NS_FD, &netns_fd, sizeof (netns_fd), err);
+  if (UNLIKELY (ret < 0))
+    return ret;
+
+  ret = append_rtattr (&req->nlh, buffer_size, IFLA_IFNAME, newifname, strlen (newifname) + 1, err);
+  if (UNLIKELY (ret < 0))
+    return ret;
+
+  ret = send_request (sock, req, err);
+  if (UNLIKELY (ret < 0))
+    return ret;
+
+  return wait_for_ack (sock, seq, buffer, buffer_size, err);
+}
+
+int
+move_network_device (const char *ifname, const char *newifname, int netns_fd, libcrun_error_t *err)
+{
+  const size_t buffer_size = 8192;
+  cleanup_free char *ips = NULL;
+  cleanup_free char *buffer = xmalloc (buffer_size);
+  int ifindex;
+  int ret;
+  cleanup_close int sock = -1;
+  int wait_status;
+  pid_t pid;
+
+  sock = open_netlink_fd (err);
+  if (sock < 0)
+    return sock;
+
+  ifindex = name_to_index (sock, ifname, buffer, buffer_size, err);
+  if (UNLIKELY (ifindex < 0))
+    return ifindex;
+
+  ret = get_ip_addresses (sock, ifindex, &ips, buffer, buffer_size, err);
+  if (UNLIKELY (ret < 0))
+    return ret;
+
+  /* Bring device down.  */
+  ret = request_set_link_state_and_wait (sock, buffer, buffer_size, ifindex, false, err);
+  if (UNLIKELY (ret < 0))
+    return ret;
```

Comment:
Is there a need to do that kernel wise? AFAICT the device is moved to down in the new namespace no matter what? So I wonder if we could save one netlink roundtrip here.

Reply:
good point, I agree this can be removed